### DISCUSSION TITLE
Future idea: Add mining machine rental and paid in COFFEE?

### DISCUSSION THREAD

**Discussion by kevingzhang** [tearust/teaproject#25] [Ideas]
This is not a feature we will add to epoch2.1. Just an idea for future 2.2+ for brainstorming.
In real life, mining machine doesn't come for free. Miners either buy or rent and paying fiat such as USD.

In epoch 2.1 we just generous give miners for free due to easy start-up. In the next contest, I am thinking to add a rental fee.

There will be 3 price tiers for type A/B/C CML. The price could be 4:2:1 as the same as their performance rate.
The rental needs to be paid in COFFEE, not TEA. This makes there is a chance for people to convert TEA to COFFEE. So the exchange works like real.

Let's brainstorm on this.

**Comment by geopor** [comment]
Suggestion: stay as close to "real life" as possible.

- Since first cloud miners likely to be AWS Nitro, could use their cheapest instance **m5d.large** to base pricing on (currently .113/hour)
- Shouldn't A/B/C cost the same to mine in real life? Luck of the draw that someone gets productive seeds. 
- Long-term, I think it would be a helpful on-boarding feature to allow people to put in AWS API credentials and be able to spin up Nitro instance with our custom miner distro. For testing purposes, AWS has [mock integration using the API Gateway console](https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-mock-integration-console.html) which doesn't charge any money.

**Reply by kevingzhang** [reply]
> Since first cloud miners likely to be AWS Nitro, could use their cheapest instance m5d.large to base pricing on (currently .113/hour)

This is doable when we know how much COFFEE is one US Dollar. if we define one COFFEE is one US Dollar, but how could US Dollar has such high compound interest rate? I hope not to introduce another currency USD into the simulator mining contest game. 

One thing we can do is to have the init COFFEE equals to one USD and use the 0.113 Hour as the start price. What about this number? @geopor 

> Shouldn't A/B/C cost the same to mine in real life? Luck of the draw that someone gets productive seeds.

The goal of design A/B/C types of CML is that in real life there are different kind of mining machine. Some of them are pretty expensive, such as AWS Nitro, some of them are pretty cheap, such as RPi or custom built mining machine. We need to have some kind of constraint to diversify the eco system. Otherwise, one type of CML will rule the whole world cause a new kind of centralization. Also exposure of vulnerabilities on one single type of CML may cause the disaster to the whole network. In the future there might be more types of CML for different kind of hardware platform. To make them coexist and compete, the price could be the best solution.
 
> Long-term, I think it would be a helpful on-boarding feature to allow people to put in AWS API credentials and be able to spin up Nitro instance with our custom miner distro. For testing purposes, AWS has mock integration using the API Gateway console which doesn't charge any money.

Yes this is doable and what we plan to do. Just not in our half year plan yet.

**Comment by kevingzhang** [comment]
@liyangwood suggests we require a one-time payment when plant. more like buy rather than rent. reasons are:
- easier implement. No need to deal with the case that balance over due
- Give a chance for the miners to make decisions between stake and mine. If the mining machine is not cheap, we encourage miners to stake
- In real life, buy is probably more common than rent

**Reply by leeyemi** [reply]
@kevingzhang 
I hope you don't mind to spare you time explaining to noob like me 

1. CML is the identity of miners hardware right in real life, not simulation. Now assuming miners A has expensive hardware and got CML A and miner C hardware is CML C, I need more explanation to understand what we discuss here. Should it be better for miner A to mine instead of staking?
2. Is there any different in term of CML A performance status between mining and staking? Understand that this number will improve or decrease according performance of their hardware.
3. If miner A is mining and miner C decide to stake on miner A, in a real life if we are talking of GPU miners, miner C lending the hash power to miner A. Is my understanding is correct?
4. There are two types of staking, using CML and using TEA. Using the example on point 3 what benefit the miner A will get from other miner staking on his slot? If the concept is like borrowing hash power then surely this will benefit everyone.
5. Will CML A (miner) and CML C (staking) performance rate increase based on the scenario point 3?
6. What different does it make to miner A between CML staked or TEA staked on his slot?
7. What different does it make to miner A between CML A, stake on his slot 1 vs. CML B staked on his slot 1, or CML C staked on his slot 1?
8. if miner C already had his machine (CML 3), what makes him not to mine? If he plans not to mine shouldn't it be better for him to buy TEA and stake on other miner? Is it because of staking CML C to miner with CML A on slot 1 earning more than running his machine by himself?
9. If miner C (CMLS 3) decide to mine and there miner A (CML A) stake on miner C, Is it a waste of investment to miner A?

**Reply by kevingzhang** [reply]
> CML is the identity of miners hardware right in real life, not simulation. Now assuming miners A has expensive hardware and got CML A and miner C hardware is CML C, I need more explanation to understand what we discuss here. Should it be better for miner A to mine instead of staking?

In the contest, we did not require miners to have a real hardware to host CML. But in the real-life, there is a hardware requirement. So you own a CML doesn't mean you can mine TEA token from the thin air. You will need to buy or rent a machine, run the shell command (like the green text when you plant a CML) to start the mining machine. 

CML itself is software license. 

If you have a CML, but you do not have enough money to buy a hardware to plant it, now you have the choice to stake now, or make more money to buy a hardware to plant in the future. this is why we discussion here

**Reply by kevingzhang** [reply]
> Is there any different in term of CML A performance status between mining and staking? Understand that this number will improve or decrease according performance of their hardware.

Mining means you occupy the index zero staking slot, which has the largest weight. When you stake, you will get a higher index slot which is smaller weight cause lower income. When staking, you do not need to buy a machine, just stake on others existing mining CML. 

You can see the weight here:
![image](https://user-images.githubusercontent.com/1761809/129842963-e4825a9b-ff73-47ec-8cb3-282664b03692.png)

**Reply by kevingzhang** [reply]
> 3\. If miner A is mining and miner C decide to stake on miner A, in a real life if we are talking of GPU miners, miner C lending the hash power to miner A. Is my understanding is correct?

No, the hash power would not be lending to anyone else. It is just how we calculate the share to the miners and stakers.
When miner C has a CML c stake to miner A 's machine a. CML c is not actually mining. It is used as an investor. That is why we call it stake. 

The higher stake a CML owns (more people invest to you), the higher chance you can get real task in the competition (in real life). So get others to stake on you is a good thing on this side. 

On another side, staking on you means take away some revenue from you. 

Just like you running a company, you want others to invest to your company, of course, they own a share of your company.

**Reply by kevingzhang** [reply]
> 4\. There are two types of staking, using CML and using TEA. Using the example on point 3 what benefit the miner A will get from other miner staking on his slot? If the concept is like borrowing hash power then surely this will benefit everyone.

I explained in above 3. 
the benefit is your stakes increase your bargain power when competing with other CML to get client's tasks. This is very important in real life. Clients usually prefer to do business with larger company because they have larger stake

**Reply by kevingzhang** [reply]
> 5\. Will CML A (miner) and CML C (staking) performance rate increase based on the scenario point 3?

Yes, the higher stake you are, the more income you get. Assume A has 1 stake slots, B has 2 stake slots. Both A and B are the current performance rate, then B has double income.

**Reply by kevingzhang** [reply]
> 6\. What different does it make to miner A between CML staked or TEA staked on his slot?

No difference. as long as it is in a slot, it is the same

**Reply by kevingzhang** [reply]
> 7\. What different does it make to miner A between CML A, stake on his slot 1 vs. CML B staked on his slot 1, or CML C staked on his slot 1?


type A CML occupies 3 slots, type B CML occupies 2 slots, but type C CML only occupy one slot. That also means Type A CML worths 3000T when staking

**Reply by kevingzhang** [reply]
> 8\. if miner C already had his machine (CML 3), what makes him not to mine? If he plans not to mine shouldn't it be better for him to buy TEA and stake on other miner? Is it because of staking CML C to miner with CML A on slot 1 earning more than running his machine by himself?

I do not know. this is exactly what we want to know during the contest. It is just the same question "should I invest to Tesla or BTC". It depends on too many factors

**Reply by kevingzhang** [reply]
> 9\. If miner C (CMLS 3) decide to mine and there miner A (CML A) stake on miner C, Is it a waste of investment to miner A?

sorry, I did not get your question. can you explain in detail?

**Reply by leeyemi** [reply]
Okay now I understand CML is a license that miners need to purchase before investing in hardware specified by the license. 
What I thought previously was the miners need to purchase a hardware and in order to mine they need to get license which assigned based on their hardware specifications.


_The higher stake a CML owns (more people invest to you), the higher chance you can get real task in the competition (in real life). So get others to stake on you is a good thing on this side._ Yes this is what I expect, understand the earning will be shared, like you said the miner will have better chance to be chosen to do the tasks. (more tasks done more earnings) don't mine to share. (off topic from simulation: I did try to stake all my CML to one CML A mining and did not see the increase of earnings if more stakes increase wining of computation tasks).

 _If miner C (CMLS 3) decide to mine and there miner A (CML A) stake on miner C, Is it a waste of investment to miner A?_
No worries you had explained it CML A worth 3 slots and most probably may boost the mining power.

To differentiate the price of CML A/B/C to me is to understand what will be the ROI and how long is the payback period.

**Comment by geopor** [comment]
> 8. if miner C already had his machine (CML 3), what makes him not to mine? If he plans not to mine shouldn't it be better for him to buy TEA and stake on other miner? Is it because of staking CML C to miner with CML A on slot 1 earning more than running his machine by himself?

When deciding whether to mine with a C seed or just stake it to a more productive machine like in your example a machine staked with an A seed, you have to look at the general productivity of machines given their seed strength relative to the number of staking slots occupied on it. On average, these are the productivity numbers (at their peak) for the classes of seeds:

- **A** seeds have around 40,000 productivity
- **C** seeds have about 10,000 productivity

So if you are the miner of a C-seed machine with no one interested in staking in your comparatively less productive machine, you'll earn at 10,000 productivity level at its peak.

But taking the same C-seed and staking it on an A machine can be just as profitable, as long as there are not too many stakers. Here is an A machine with 3 total stakers. Note that if you were the slot-1 staker earning around 24%, it would be earning approximately the same as the miner earning 100% of their C-seed mining machine.

<img width="1136" alt="Screen Shot 2021-08-17 at 11 38 31 PM" src="https://user-images.githubusercontent.com/86096370/129849994-d29fbf9c-37b5-4684-b77b-9eea7e8e752f.png">

Holding that stake-1 slot is fine with up to 3 stakers, but once the fourth one comes along you start to dip around 20% and then you're earning less than if you had mined with the C seed. But a strategy that can be used is to stake your C seeds in more powerful machines until too many people jump on board. Then the seeds can be unstaked and planted into their own mining machines.

>9. If miner C (CMLS 3) decide to mine and there miner A (CML A) stake on miner C, Is it a waste of investment to miner A?

Yes. The A seed is something special for mining machines and would be wasted if placed as a stake in a C-machine, or even an A-machine.

**Reply by kevingzhang** [reply]
In the next epoch, each type of seed can only be planted to the corresponding type of machine. So planting a type C seed to a type A machine and get the type A performance is not possible. Even you have enough fund to buy a type A machine, but planting a type C seed to it won't work.
